vllm/core/block_manager_v2.py

```diff
@@ -116,18 +116,39 @@ def allocate(self, seq_group: SequenceGroup) -> None:
         for seq in waiting_seqs[1:]:
             self.block_tables[seq.seq_id] = block_table.fork()
 
-    def can_append_slot(self, seq_group: SequenceGroup) -> bool:
-        # Simple heuristic: If there is at least one free block
-        # for each sequence, we can append.
+    def can_append_slots(self, seq_group: SequenceGroup,
+                         num_lookahead_slots: int) -> bool:
+        """Determine if there is enough space in the GPU KV cache to continue
+        generation of the specified sequence group.
+
+        We use a worst-case heuristic: assume each touched block will require a
+        new allocation (either via CoW or new block). We can append slots if the
+        number of touched blocks is less than the number of free blocks.
+
+        "Lookahead slots" are slots that are allocated in addition to the slots
+        for known tokens. This is used by speculative decoding when speculating
+        future tokens.
+        """
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.RUNNING):
+            block_table = self.block_tables[seq.seq_id]
+            num_new_tokens = seq.get_len() - block_table.num_full_slots
+
+            num_touched_blocks += (
+                block_table.get_num_blocks_touched_by_new_tokens(
+                    # NOTE: we treat lookahead slots as new tokens for the
+                    # worst-case estimation.
+                    num_new_tokens + num_lookahead_slots))
+
         num_free_gpu_blocks = self.block_allocator.get_num_free_blocks(
             Device.GPU)
-        num_seqs = seq_group.num_seqs(status=SequenceStatus.RUNNING)
-        return num_seqs <= num_free_gpu_blocks
+        return num_touched_blocks <= num_free_gpu_blocks
 
-    def append_slot(
+    def append_slots(
         self,
         seq: Sequence,
-    ) -> Optional[Tuple[int, int]]:
+        num_lookahead_slots: int,
```

Comment:
`num_lookahead_slots` is not used in this function, is it expetced?

Reply:
Not expected, will add `ensure_num_empty_slots` call with `num_lookahead_slots` + see why the test didn't catch this.

Reply:
fixed; added test.